dex-it/src/test/scala/com/wavesplatform/it/sync/MarketOrderTestSuite.scala

```diff
@@ -1,118 +1,449 @@
-    pending
+  "Validation market orders" - {
+    val price = 0.4.usd
+
+    "should be accepted if there is no way to fill it completely (sum of all orders in order book < amount of market order" in {
+      placeOrders(bob, wavesUsdPair, SELL)(100.waves -> price)
+
+      val total = node.orderBook(wavesUsdPair).bids.map(_.amount).sum
+
+      node.waitOrderProcessed(wavesUsdPair,
+                              node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, BUY, total + 1000.waves, price, fixedFee)).message.id)
+    }
+
+    "should be accepted if user doesn't have enough Waves to pay fee, but he take waves from order result " in {
+      val amount   = 100.waves
+      val price    = 0.1.usd
+      val transfer = 100.usd
+
+      val account = createAccountWithBalance {
+        transfer -> Some(UsdId.toString)
+      }
+
+      placeOrders(alice, wavesUsdPair, SELL)(
+        amount -> price
+      )
+
+      node.waitOrderProcessed(wavesUsdPair, node.placeMarketOrder(node.prepareOrder(account, wavesUsdPair, BUY, amount, price, fixedFee)).message.id)
+
+      node.accountBalances(account.toAddress.toString)._1 should be(amount - fixedFee)
+    }
+
+    "should be rejected if the price is too low for completely filling by current opened orders (BUY)" in {
+      val amount      = 100.waves
+      val marketPrice = price - 0.1.usd
+
+      placeOrders(bob, wavesUsdPair, SELL)(amount -> price)
+
+      assertBadRequestAndMessage(node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, BUY, amount, marketPrice, fixedFee)),
+                                 tooLowPrice("buy", "0.3"))
+    }
+
+    "should be rejected if the price is too high for completely filling by current opened orders (SELL)" in {
+      val amount      = 100.waves
+      val marketPrice = price + 0.1.usd
+
+      placeOrders(bob, wavesUsdPair, OrderType.BUY)(amount -> price)
+
+      assertBadRequestAndMessage(node.placeMarketOrder(node.prepareOrder(alice, wavesUsdPair, OrderType.SELL, amount, marketPrice, fixedFee)),
+                                 tooHighPrice("sell", "0.5"))
+    }
+
+    "should be rejected if amount of the buy market order more then user could buy" ignore /* because of TODO: DEX-457 */ {
+      val amount   = 101.waves
+      val price    = 1.1.usd
+      val transfer = 100.usd
+
+      val account = createAccountWithBalance {
+        transfer -> Some(UsdId.toString)
+      }
+
+      placeOrders(alice, wavesUsdPair, SELL)(
+        amount -> price
+      )
+
+      assertBadRequestAndMessage(
+        node.placeMarketOrder(node.prepareOrder(account, wavesUsdPair, BUY, amount, price, fixedFee)),
+        s"Not enough tradable balance. The order requires 0 WAVES and 111.1 ${UsdId}, but available are 111.1 ${UsdId} and 0 WAVES"
+      )
+
+    }
+
+    "should be rejected if user has enough balance to fill market order, but has not enough balance to pay fee in another asset" ignore /* because of TODO: DEX-457 */ {
```

Comment:
See [comments](https://jira.wavesplatform.com/browse/DEX-457?focusedCommentId=58977&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-58977) to DEX-457 issue

Reply:
ok